xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -919,6 +919,9 @@ private static String validateRoutes(List<EnvoyProtoData.Route> routes) {
       if (!routeMatch.isDefaultMatcher()) {
         return "The last route must be the default route";
       }
+      if (!routeMatch.isCaseSensitive()) {
```

Comment:
Data validation should be done against the originally received proto messages at first place instead of against converted data. Conversion should not happen if the original proto message contains invalid data.

Reply:
But how? I don't think it's so practical. The conversion is just pure data transformation from raw proto class to EnvoyProtoData class, it has no business logic, so it's transparent to validation, you can either validate raw proto with proto getters, or validate transformed data with EnvoyProtoData getters.

Reply:
The conversion itself is (straightforward) business logic (it interprets the proto messages). 

Whatever, your previous change already shifted things towards that direction (e.g., `findRoutesInRouteConfig(...)` method does some validation against `RouteConfiguration` proto, then it does the conversion for `Route`, then do more validation on converted data), which is not a big deal in terms of correctness but I would just consider it bad logic style.

Reply:
>your previous change already shifted things towards that direction (e.g., findRoutesInRouteConfig(...) method does some validation against RouteConfiguration proto,

You don't have  the `EnvoyProtoData` counterpart of `RouteConfiguration` proto, so in `findRoutesInRouteConfig` it just used raw proto getters directly at the `RouteConfiguration` level. Any validation or business logic are at that level. It does not use any `Route` level proto getters. If it had `EnvoyProtoData` counterpart, very likely we might converted first and then do validation.

`caseSensitive` is like `hasRegex`, grpc just chose not to support them or currently not to support them. This is the business logic that the conversion does not have to know. The conversion just blindly feed all the data the business logic needs, as how the logic will process the data, support it or not, we don't care before the conversion.

Reply:
Another aspect of my thought is, the converted data is something that gRPC would use in gRPC's internal logic such LB policies. So it should only contain fields that gRPC is interested in. Since responses with case-sensitive routes are rejected at the first place before the information is passed to gRPC's internal, this case-sensitivity information would not be needed and the converted data should not have it.

I would consider this as a style problem, what I am worrying is we would be frustrated for maintainability as things like this spread more and more.

Reply:
> the converted data is something that gRPC would use in gRPC's internal logic such LB policies

>I would consider this as a style problem

I'm thinking about how the style problem comes. The LB policies use `ConfigUpdate` and `EnpointUpdate` etc, and they in turn happen to use types in `EnovyProtoData` because some sub-structures of `ConfigUpdate` and `EnpointUpdate` just currently (more or less) keep the proto structure. I think `ConfigUpdate` and `EnpointUpdate` shouldn't really use any type obtained from `EnovyProtoData.fromEnvoyProto*()` for sub-structures, they are causing the trouble and might be breaking easily.

Reply:
What's the problem of converted data keeping the proto structure? 

My thought about this issue is:
`EnovyProtoData.fromEnvoyProto*()` is an easy/lazy way to quickly convert a proto message to POJO and preserve only fields that we are interested in. It is great for converting proto messages that we do not care its data. The problem is, in many cases, we need go into the proto message to validate its field values and then call `EnovyProtoData.fromEnvoyProto*()` to convert it. The two approaches we are discussing are:

- Validate data on proto message then convert to POJO
   - pros:
      - Clean definitions for converted objects, having only fields that gRPC is interested in. Every instantiated object is meaningful and usable for gRPC.

   - cons:
      - Validating (with business logic) the proto message requires walking through the message, converting it to POJO requires walking through it again if using `EnovyProtoData.fromEnvoyProto*()`. (one solution is to build the converted object step by step as the validation goes, instead of using `EnovyProtoData.fromEnvoyProto*()`. concern: what if the proto message is too complicated? That would require a lot of setters.).


- Convert proto to POJO then validate data on converted object
   - pros:
      - Conversion is easy and mechanical.

   - cons:
      - Would need to preserve fields in proto messages that gRPC do not use, as we may have validation rules for them. For example, `case_sensitive` field in `RouteMatch` and unknown fields that we may validate (design under discussion?).


I am thinking if we could move all the validation business logic into converters, so that we take advantage of both approaches above. It would be each converted type has a `EnovyProtoData.fromEnvoyProto*()` method, validations are performed inside it. If any rule is violated, throw some exception containing the error message (or we could do a similar trick as `ConfigOrError`).

Reply:
The problem is that the design keeps changing. Validation also keeps changing. It might be a pain to incorporate new change. 

I'm okay with `EnovyProtoData.fromEnvoyProto*()` throws Exception approach.

Reply:
Sure, we can extend our discussion regarding response data validation and conversion further after this PR.